### PR TITLE
GH-118093: Better handling of short and mid-loop traces

### PR DESCRIPTION
**Pull request by brandtbucher**
This makes two changes to trace projection:
- Handle a single backward jump *anywhere* in a trace, rather than just at the beginning.
- Rather than having a minimum trace length, just ensure that at least one instruction is fully translated. This prevents us from requiring the translation of several tiny instructions or "successfully" half-translating a large instruction.

Perf is [neutral](https://github.com/faster-cpython/benchmarking-public/blob/main/results/bm-20240724-3.14.0a0-f1f5d13-JIT/bm-20240724-linux-x86_64-brandtbucher-tracing_jumps-3.14.0a0-f1f5d13-vs-base.svg), but stats [move in the right direction](https://github.com/faster-cpython/benchmarking-public/blob/main/results/bm-20240724-3.14.0a0-f1f5d13-JIT/bm-20240724-azure-x86_64-brandtbucher-tracing_jumps-3.14.0a0-f1f5d13-pystats-vs-base.md): 83% reduction in "inner loop found", 7% reduction in "trace too short", and 5% fewer optimization attempts overall. We execute 0.5% more uops and 0.5% fewer traces. 2% fewer tier one instructions executed overall.

* Issue: gh-118093

### CODE REVIEW

File: Objects/genobject.c
Comment (markshannon) [inline Objects/genobject.c:345]: Why have these asserts been removed?

File: Python/optimizer.c
Comment (markshannon) [inline Python/optimizer.c]: If we've hit a stack underflow, it means that we are stopping projection, no that we've made no progress.
It would imply that we've already made plenty of progress.
Maybe add an `overflow` flag and check that at the end?

**Review by markshannon — COMMENTED** [review]
A couple of questions

File: Objects/genobject.c
Comment (brandtbucher) [reply Objects/genobject.c:345]: The old code was buggy... it couldn't handle an `ENTER_EXECUTOR` sitting on top of the `RESUME`.

File: Python/optimizer.c
Comment (brandtbucher) [reply Python/optimizer.c]: Sure. This just felt like a faithful translation of the original code.

I'm a bit wary of adding a bunch of flags that are all essentially doing the same thing, but one for this probably shouldn't be too bad (and would fix the weird behavior of our stats where "underflow" failures are double-counted as "too short" failures as well).

Comment (brandtbucher) [reply Python/optimizer.c]: I realized that we can just replace this with `return 0;`.

File: Objects/genobject.c
Comment (markshannon) [reply Objects/genobject.c:345]: That makes `is_resume` buggy, not the assert, does it not? I think it would be better to fix `is_resume` to account for `ENTER_EXECUTOR`.

Similarly for the following assert.

Comment (brandtbucher) [reply Objects/genobject.c:345]: I guess I *can*, but I was just hesitant to because this `is_resume` function is used during generator finalization for an optimized fast path and I didn't want to slow it down too much with `ENTER_EXECUTOR` handling (which requires quite a bit of fishing around in the code object).

It's a fairly straightforward change to make if you think it's a good idea, though. Your call.

Comment (brandtbucher) [reply Objects/genobject.c:345]: I went ahead and fixed `is_resume`.

File: Objects/genobject.c
Comment (markshannon) [inline Objects/genobject.c]:
```suggestion
    int offset = (int)(frame->instr_ptr - _PyCode_CODE(code));
```

File: Objects/genobject.c
Comment (markshannon) [inline Objects/genobject.c:345]: The new implementation of `is_resume` is a bit strange, with it returning the `oparg` via an output parameter.
I think we need two functions. One that *quickly* returns yes/probably not for the gen_close and another one for sanity checking that can be slow. For that we will want a better API for getting the original instruction.

That's all too much for this PR though. So, can you go back to to just dropping the asserts?
I've made an issue for sorting out `is_resume` and a better API for getting the original instruction. https://github.com/python/cpython/issues/122390

**Review by markshannon — CHANGES_REQUESTED** [review]
Can you:
* Go back to removing the asserts, and leave `is_resume` alone.
* Add a comment linking to https://github.com/python/cpython/issues/122390

Consider it approved once you've done that

**Pull request merged**
